packages/gatsby/src/schema/__tests__/node-model.js

```diff
@@ -0,0 +1,446 @@
+      const type = `Post`
+      const query = { filter: { frontmatter: { published: { eq: false } } } }
+      const firstOnly = true
+      const result = await nodeModel.runQuery({ query, firstOnly, type })
+      expect(result.id).toBe(`post1`)
+    })
+
+    it(`returns all results`, async () => {
+      const type = `Post`
+      const query = { filter: { frontmatter: { published: { eq: false } } } }
+      const firstOnly = false
+      const result = await nodeModel.runQuery({ query, firstOnly, type })
+      expect(result.length).toBe(2)
+      expect(result[0].id).toBe(`post1`)
+      expect(result[1].id).toBe(`post3`)
+    })
+
+    it(`creates page dependencies`, async () => {
+      const type = `Post`
+      const query = { filter: { frontmatter: { published: { eq: false } } } }
+      const firstOnly = false
+      await nodeModel.runQuery({ query, firstOnly, type }, { path: `/` })
+      expect(createPageDependency).toHaveBeenCalledTimes(2)
+      expect(createPageDependency).toHaveBeenCalledWith({
+        path: `/`,
+        nodeId: `post1`,
+      })
+      expect(createPageDependency).toHaveBeenCalledWith({
+        path: `/`,
+        nodeId: `post3`,
+      })
+    })
+
+    it(`creates page dependencies with connection type`, async () => {
+      const type = `Post`
+      const query = { filter: { frontmatter: { published: { eq: false } } } }
+      const firstOnly = false
+      await nodeModel.runQuery(
+        { query, firstOnly, type },
+        { path: `/`, connectionType: `Post` }
+      )
+      expect(createPageDependency).toHaveBeenCalledTimes(1)
+      expect(createPageDependency).toHaveBeenCalledWith({
+        path: `/`,
+        connection: `Post`,
+      })
+    })
+
+    it(`doesn't allow querying union types`, () => {
+      const type = `AllFiles`
+      const query = {}
+      const firstOnly = true
+      const result = nodeModel.runQuery({ query, firstOnly, type })
+      return expect(result).rejects.toThrowError(
+        `Querying GraphQLUnion types is not supported.`
+      )
+    })
+
+    it(`handles interface types`, async () => {
+      const type = `TeamMember`
+      const query = { name: { ne: null } }
+      const firstOnly = true
+      const result = await nodeModel.runQuery({ query, firstOnly, type })
+      expect(result.name).toBe(`Person1`)
+    })
+
+    it(`allows passing GraphQLType instead of type name`, async () => {
+      const type = schema.getType(`File`)
+      const query = {
+        filter: {
+          children: { elemMatch: { internal: { type: { eq: `Post` } } } },
+        },
+      }
+      const firstOnly = false
+      const result = await nodeModel.runQuery({ query, firstOnly, type })
+      expect(result.length).toBe(2)
+      expect(result[0].id).toBe(`file1`)
+      expect(result[1].id).toBe(`file3`)
+    })
+  })
+
+  describe(`findRootNodeAncestor`, () => {
+    it(`returns an object's top most ancestor node`, () => {
```

Comment:
I'm not 100% clear why finding the top most ancestor is default, as opposed to finding an object's parent node (which is what you'd need most of the time)?